Fetch AP HT/VHT capabilities via nl80211

mariomaz opened the pull request:
See #707

mariomaz committed:
bcl: add HT_MCS_SET_SIZE and VHT_MCS_SET_SIZE consts

Add HT_MCS_SET_SIZE and VHT_MCS_SET_SIZE to be used across the whole
project.

The values for these macros are described in <linux/nl80211.h> to be
16 and 32 bytes respectively.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bwl: use HT_MCS_SET_SIZE and VHT_MCS_SET_SIZE consts

Update nl80211_client interface to use HT_MCS_SET_SIZE and
VHT_MCS_SET_SIZE consts instead of literal values.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: add AP HT/VHT Capability fields

Add new AP HT/VHT Capability fields to struct sNodeHostap so they can
be sent from ap_manager_thread to slave_thread inside
ACTION_APMANAGER_JOINED_NOTIFICATION.

Add those fields to message cACTION_BACKHAUL_ENABLE too so they can be
sent from slave_thread to backhaul_manager.

backhaul_manager stores those fields into m_radio_info_map to later
build AP Capabilities TLVs.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
tlvf: update autogenerated files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: send AP HT/VHT Capability fields from bwl to backhaul manager

Reading radio info should only be done by the BWL backend. We read from
hostapd whatever we can, and supplement what's missing with the
nl80211_client.

Backhaul manager needs AP HT/VHT Capability data to fill in some TLVs in
AP Capability Report message. This data must be sent from
ap_manager_thread to slave_thread inside
cACTION_APMANAGER_JOINED_NOTIFICATION and then to backhaul_manager
inside cACTION_BACKHAUL_ENABLE where are saved into m_radio_info_map.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

tomereli commented on common/beerocks/bwl/include/bwl/nl80211_client.h:
```diff
@@ -24,15 +24,15 @@
  * the MCS set as defined in 802.11n
  */
-static constexpr size_t ht_mcs_set_size = 16;
+static constexpr size_t ht_mcs_set_size = beerocks::message::HT_MCS_SET_SIZE;
```
I think this can be squashed to the previous commit (nitpick)

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -61,6 +61,14 @@ cACTION_BACKHAUL_ENABLE:
+  ht_capability: uint16_t
+  ht_mcs_set:
+    _type: uint8_t
+    _length: [ "beerocks::message::HT_MCS_SET_SIZE" ]
```
```suggestion
    _length: [ beerocks::message::HT_MCS_SET_SIZE ]
```
Quoting is not needed.

mariomaz replied:
I get the following error if not quoting:
```
-- Generating Beerocks TLVF files...
Traceback (most recent call last):
  File "/home/developer/workspace/framework/tlvf/tlvf.py", line 1999, in <module>
    main()
  File "/home/developer/workspace/framework/tlvf/tlvf.py", line 1996, in main
    tlvf = TlvF(args.src_path, args.yaml_path, args.out_path, args.conf, args.print_dependencies, args.print_outputs)
  File "/home/developer/workspace/framework/tlvf/tlvf.py", line 404, in __init__
    self.loadAllYamlFilesToDB()
  File "/home/developer/workspace/framework/tlvf/tlvf.py", line 1799, in loadAllYamlFilesToDB
    yaml_inst = self.loadYaml(fname)
  File "/home/developer/workspace/framework/tlvf/tlvf.py", line 1919, in loadYaml
    raise Exception("  %s" % str(exc))
Exception:   while scanning a plain scalar
  in "/home/developer/workspace/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml", line 67, column 16
found unexpected ':'
  in "/home/developer/workspace/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml", line 67, column 24
Please check http://pyyaml.org/wiki/YAMLColonInFlowContext for details.
CMake Error at common/beerocks/tlvf/CMakeLists.txt:28 (message):
  -- /home/developer/workspace/framework/tlvf/tlvf.py --print-outputs failed!
```

tomereli replied:
@LiorAmram please comment - I think its a bug in `tlvf.py` which doesn't generate this properly

rmelotte replied:
From what I see in other yaml files it looks like length variables are unquoted when it's defined in the yaml file itself, while they are quoted when they are defined in header (C++) files.

I wonder why the length is always a one-element (or empty) list though.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -61,6 +61,14 @@ cACTION_BACKHAUL_ENABLE:
+  vht_capability: uint32_t
+  vht_mcs_set:
+    _type: uint8_t
+    _length: [ "beerocks::message::VHT_MCS_SET_SIZE" ]
```
```suggestion
    _length: [ beerocks::message::VHT_MCS_SET_SIZE ]
```

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -267,6 +267,14 @@ sNodeHostap:
+  ht_capability: uint16_t
+  ht_mcs_set:
+    _type: uint8_t
+    _length: [ "beerocks::message::HT_MCS_SET_SIZE" ]
```
```suggestion
    _length: [ beerocks::message::HT_MCS_SET_SIZE ]
```

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -267,6 +267,14 @@ sNodeHostap:
+  vht_capability: uint32_t
+  vht_mcs_set:
+    _type: uint8_t
+    _length: [ "beerocks::message::VHT_MCS_SET_SIZE" ]
```
```suggestion
    _length: [ beerocks::message::VHT_MCS_SET_SIZE ]
```

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -270,7 +270,13 @@ class backhaul_manager : public btl::transport_socket_thread {
             beerocks::eWiFiBandwidth::BANDWIDTH_UNKNOWN; /**< Maximum supported bandwidth */
-        beerocks_message::sVapsList vaps_list;           /**< List of VAPs in radio. */
+        uint16_t ht_capability = 0;                      /**< HT capabilities */
+        std::basic_string<uint8_t>
```
Why use basic_string and not `std::array`? Isn't it always 16 bytes?

mariomaz replied:
Changed to std::array

tomereli replied:
Actually now @arnout introduced a new class `bytes` so perhaps its better to use this one instead - see dev/client-caps-report-assoc_req-with-bytes (nitpick)

tomereli reviewed (approved):
I just don't understand why we are using `basic_string` for describing fixed sized data structures. No comments other than that so approving.

mariomaz committed:
bwl: add flags for supported HT/VHT/HE capabilities

Add boolean flags to signal when each of the HT, VHT or HE
capabilities are supported or not.

These flags will later be used to decide if corresponding TLVs must be
added or not to the AP capabilities response message.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>
